File: src/main/scala/kim/dencka/easy/other/FirstBadVersion.scala
package kim.dencka.easy.other

import kim.dencka.easy.other.FirstBadVersion.isBadVersion

// KIM_INFO IT SOLVED

class FirstBadVersion extends App {
  def firstBadVersion(n: Int): Int = {
    @annotation.tailrec
    def run(low: Int, high: Int): Int = {
      val mid = low + (high - low) / 2
      if (!isBadVersion(mid - 1) && isBadVersion(mid)) mid
      else if (isBadVersion(mid)) run(low, mid - 1)
      else run(mid + 1, high)
    }
    run(1, n)
  }
}

object FirstBadVersion {
  def isBadVersion(version: Int): Boolean = ???
}

File: src/main/scala/kim/dencka/java/easy/other/FirstBadVersion.java
package kim.dencka.java.easy.other;

class FirstBadVersion {
    public static void main(String[] args) {
        System.out.println(firstBadVersion(5));
    }

    public static int firstBadVersion(int n) {
        int low = 0, high = n, mid;
        int result = -1;
        while (low <= high) {
            mid = low + (high - low) / 2;
            if (isBadVersion(mid)) {
                result = mid;
                high = mid - 1;
            } else low = mid + 1;
        }
        return result;
    }


    public static boolean isBadVersion(int version) {
        System.out.println(version);
        return false;
    }

}
